llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp

```diff
@@ -6855,6 +6855,47 @@ Instruction *InstCombinerImpl::visitICmpInst(ICmpInst &I) {
   if (Value *V = simplifyICmpInst(I.getPredicate(), Op0, Op1, Q))
     return replaceInstUsesWith(I, V);
 
+  {
+    Value *X;
+    const APInt *C1, *C2;
+    ICmpInst::Predicate Pred;
+    if ((match(&I, m_ICmp(Pred, m_SRem(m_Value(X), m_NonNegative(C1)),
+                          m_APInt(C2))) &&
+         *C2 == *C1 - 1 &&
+         (Pred == ICmpInst::ICMP_SLT || Pred == ICmpInst::ICMP_SGE)) ||
+        (match(&I,
+               m_ICmp(Pred, m_SRem(m_Value(X), m_Negative(C1)), m_APInt(C2))) &&
+         *C2 == *C1 + 1 &&
+         (Pred == ICmpInst::ICMP_SGT || Pred == ICmpInst::ICMP_SLE))) {
+      // icmp slt (X s% C), (C - 1) --> icmp ne (X s% C), (C - 1), if C >= 0
+      // icmp sge (X s% C), (C - 1) --> icmp eq (X s% C), (C - 1), if C >= 0
```

Comment:
We always canonicalize `icmp sge X, C` into `icmp sgt X, C - 1`.

Reply:
So the check should be on `icmp sgt (X % C), C-2`?

Reply:
@dtcxzyw

Reply:
> So the check should be on `icmp sgt (X % C), C-2`?

Yes.